core/src/commonMain/kotlin/com/xebia/functional/xef/llm/assistants/AssistantThread.kt

```diff
@@ -72,208 +69,110 @@ class AssistantThread(
   suspend fun createRun(request: CreateRunRequest): RunObject =
     api.createRun(threadId, request, configure = ::defaultConfig).addMetrics(metric)
 
+  fun createRunStream(assistant: Assistant, request: CreateRunRequest): Flow<RunDelta> = flow {
+    api
+      .createRunStream(threadId, request, configure = ::defaultConfig)
+      .map { RunDelta.fromServerSentEvent(it) }
+      .collect { event ->
+        when (event) {
+          // submit tool outputs and join streams
+          is RunDelta.RunRequiresAction -> {
+            val steps =
+              runSteps(event.run.id).map { metric.assistantCreateRunStep(event.run.id) { it } }
+
+            steps.forEach { step ->
+              val calls = step.stepDetails.toolCalls()
+
+              val run = getRun(event.run.id)
+              if (
+                run.status == RunObject.Status.requires_action &&
+                  run.requiredAction?.type == RunObjectRequiredAction.Type.submit_tool_outputs
+              ) {
+                val callsResult: List<Pair<String, Assistant.Companion.ToolOutput>> =
+                  calls
+                    .filterIsInstance<
+                      RunStepDetailsToolCallsObjectToolCallsInner.CaseRunStepDetailsToolCallsFunctionObject
+                    >()
+                    .parMap { toolCall -> executeToolCall(toolCall, assistant) }
+                    .filterNotNull()
```

Comment:
@raulraja you added `parMapNotNull` into Arrow couple months ago 😜 We can use that here.
The suggestion will fail with a missing import though.

```suggestion
                    .parMapNotNull { toolCall -> executeToolCall(toolCall, assistant) }
```